packages/node/src/utils/kyve/kyve.ts

```diff
@@ -0,0 +1,409 @@
+
+    const kyve = new KyveApi(storageUrl, tmpCacheDir, poolId, lcdClient);
+
+    return kyve;
+  }
+
+  static async fetchPoolId(
+    chainId: string,
+    lcdClient: KyveLCDClientType,
+  ): Promise<string> {
+    const poolsResponse =
+      (await lcdClient.kyve.query.v1beta1.pools()) as unknown as QueryPoolsResponse;
+
+    for (const p of poolsResponse.pools) {
+      try {
+        const config = JSON.parse(p.data.config);
+        if (config.network === chainId) {
+          return p.id; // Return the matching pool ID
+        }
+      } catch (error) {
+        throw new Error(
+          `Error parsing JSON for pool with id ${p.id}: ${error}`,
+        );
+      }
+    }
+
+    throw new Error(`${chainId} is not available on Kyve network`);
+  }
+
+  private decodeBlock(block: JsonRpcSuccessResponse): BlockResponse {
+    return Responses.decodeBlock({
+      id: 1,
+      jsonrpc: '2.0',
+      result: block,
+    });
+  }
+
+  private decodeBlockResult(
+    blockResult: JsonRpcSuccessResponse,
+  ): BlockResultsResponse {
+    return Responses.decodeBlockResults({
+      id: 1,
+      jsonrpc: '2.0',
+      result: blockResult,
+    });
+  }
+
+  private getBundleFilePath(id: string): string {
+    return path.join(this.tmpCacheDir, `bundle_${id}.json`);
+  }
+
+  private async getBundleById(bundleId: number): Promise<BundleDetails> {
+    const bundleDetail =
+      await this.lcdClient.kyve.query.v1beta1.finalizedBundle({
+        pool_id: this.poolId,
+        id: bundleId.toString(),
+      });
+    return bundleDetail as BundleDetails;
+  }
+
+  private async getLatestBundleId(): Promise<number> {
+    return (
+      parseDecimal(
+        (
+          await this.lcdClient.kyve.query.v1beta1.finalizedBundles({
+            pool_id: this.poolId,
+            index: '1',
+            pagination: {
+              limit: '1',
+            },
+          })
+        ).pagination.total,
+      ) - 1
+    ); // bundle id starts from 0
+  }
+
+  private async getBundleId(height: number): Promise<number> {
+    const latestBundleId = await this.getLatestBundleId();
+
+    let low = this.currentBundleId;
+    let high = latestBundleId;
+    let startBundleId = -1; // Initialize to an invalid ID initially
+
+    while (low <= high) {
+      const mid = Math.floor((low + high) / 2);
+      const midBundle = await this.getBundleById(mid);
+
+      const fromKey = parseDecimal(midBundle.from_key);
+      const toKey = parseDecimal(midBundle.to_key);
+
+      if (height >= fromKey && height <= toKey) {
+        startBundleId = mid;
+        this.currentBundleId = startBundleId;
+        return startBundleId;
+      }
+
+      if (height > toKey) {
+        low = mid + 1;
+      } else {
+        high = mid - 1;
+      }
+    }
+    throw new Error(`No suitable bundle found for height ${height}}`);
+  }
+
+  private findBlockByHeight(
+    height: number,
+    fileCacheData: UnZippedKyveBlockReponse[],
+  ): UnZippedKyveBlockReponse {
```

Comment:
Also worth handling where this is called so we don't just get a `cannot read property of undefined` error

```suggestion
  ): UnZippedKyveBlockReponse | undefined {
```